semantic/src/main/kotlin/tools/samt/semantic/checks/ValidTypeExpressionCheck.kt

```diff
@@ -0,0 +1,155 @@
+
+            is GenericSpecializationNode -> diagnostics.reportError(
+                "Generic type are currently not supported",
+                expression.location
+            )
+
+            is BooleanNode,
+            is NumberNode,
+            is StringNode,
+            is ObjectNode,
+            is ArrayNode,
+            is RangeExpressionNode,
+            is WildcardNode,
+            -> diagnostics.reportError("Base type for constraint is not valid", expression.location)
+        }
+    }
+
+    private fun checkConstraintArgumentExpression(expression: ExpressionNode) {
+        when (expression) {
+            is CallExpressionNode -> {
+                if (expression.base !is BundleIdentifierNode) {
+                    diagnostics.reportError("Constraint argument name must be an identifier", expression.base.location)
+                }
+                for (argument in expression.arguments) {
+                    checkLiteralExpression(argument)
+                }
+            }
+            is RangeExpressionNode -> checkRangeExpression(expression)
+
+            is BooleanNode,
+            is NumberNode,
+            is StringNode,
+            is IdentifierNode,
+            is BundleIdentifierNode,
+            is OptionalDeclarationNode,
+            is ObjectNode,
+            is ArrayNode,
+            is WildcardNode,
+            is GenericSpecializationNode,
+            -> diagnostics.reportError("Invalid constraint, examples of valid constraints are 'size(1..*)', 'pattern(\"a-z\")' and '1..*'", expression.location)
+        }
+    }
+
+    private fun checkLiteralExpression(expression: ExpressionNode) {
+        when (expression) {
+            is BooleanNode,
+            is NumberNode,
+            is StringNode,
+            -> Unit
+
+            is ObjectNode -> expression.fields.forEach { field -> checkLiteralExpression(field.value) }
+            is ArrayNode -> expression.values.forEach { value -> checkLiteralExpression(value) }
+            is RangeExpressionNode -> checkRangeExpression(expression)
+
+            is WildcardNode -> diagnostics.reportError("Cannot use wildcard outside of ranges", expression.location)
+
+            is IdentifierNode,
+            is BundleIdentifierNode,
+            is OptionalDeclarationNode,
+            is CallExpressionNode,
+            is GenericSpecializationNode,
+            -> diagnostics.reportError("Invalid type expression", expression.location)
+        }
+    }
+
+    private fun checkRangeExpression(range: RangeExpressionNode) {
+        when {
+            range.left is FloatNode && range.right is FloatNode -> Unit
+            range.left is IntegerNode && range.right is IntegerNode -> Unit
+            range.left is WildcardNode && range.right is NumberNode -> Unit
+            range.left is NumberNode && range.right is WildcardNode -> Unit
+            else -> diagnostics.reportError("Range must be between two integers or two floats", range.location)
```

Comment:
I think Int..Float and Float..Int constraints should be allowed. It is easier to write `0..0.5` than it is `0.0..0.5`

Reply:
Fair point, will update the code and test

Reply:
We can normalize these nodes earlier / later to have both literals be of the same type